Dockerfile

```diff
@@ -19,33 +19,80 @@ ARG GOTESTSUM_VERSION=v1.9.0
 ARG GO_VERSION=1.20
 ARG ALPINE_VERSION=3.18
 
+# ALPINE_CACHE defaults to the "alpine-cache-local" stage in this image.
+# Can be overridden to a custom image for reproducible builds.
+ARG ALPINE_CACHE=alpine-cache-local
+
 # minio for s3 integration tests
 FROM minio/minio:${MINIO_VERSION} AS minio
 FROM minio/mc:${MINIO_MC_VERSION} AS minio-mc
 
 # alpine base for buildkit image
 # TODO: remove this when alpine image supports riscv64
-FROM alpine:${ALPINE_VERSION} AS alpine-amd64
-FROM alpine:${ALPINE_VERSION} AS alpine-arm
-FROM alpine:${ALPINE_VERSION} AS alpine-arm64
-FROM alpine:${ALPINE_VERSION} AS alpine-s390x
-FROM alpine:${ALPINE_VERSION} AS alpine-ppc64le
-FROM alpine:edge@sha256:2d01a16bab53a8405876cec4c27235d47455a7b72b75334c614f2fb0968b3f90 AS alpine-riscv64
+FROM --platform=amd64   alpine:${ALPINE_VERSION} AS alpine-amd64
+FROM --platform=arm     alpine:${ALPINE_VERSION} AS alpine-arm
+FROM --platform=arm64   alpine:${ALPINE_VERSION} AS alpine-arm64
+FROM --platform=s390x   alpine:${ALPINE_VERSION} AS alpine-s390x
+FROM --platform=ppc64le alpine:${ALPINE_VERSION} AS alpine-ppc64le
+FROM --platform=riscv64 alpine:edge@sha256:2d01a16bab53a8405876cec4c27235d47455a7b72b75334c614f2fb0968b3f90 AS alpine-riscv64
 FROM alpine-$TARGETARCH AS alpinebase
+FROM --platform=$BUILDPLATFORM alpine AS alpinebase-buildplatform
 
 # xx is a helper for cross-compilation
 FROM --platform=$BUILDPLATFORM tonistiigi/xx:1.2.1 AS xx
 
+FROM alpinebase-buildplatform AS alpine-cache-local-base
+RUN mkdir -p /etc/apk/cache && \
+  apk update && \
+  apk cache download \
+    bash \
+    clang curl \
+    dnsmasq \
+    file fuse \
+    git git-daemon gzip \
+    iptables ip6tables \
+    lld \
+    make \
+    musl-dev \
+    nodejs npm \
+    openssh openssh-client \
+    pkgconfig \
+    shadow shadow-uidmap sudo \
+    tar \
+    vim
+COPY --link --from=xx / /
+ARG TARGETPLATFORM
+RUN xx-apk update && \
+  xx-apk cache download \
+    fuse3 fuse-overlayfs \
+    gcc \
+    git \
+    musl-dev \
+    libseccomp-dev libseccomp-static \
+    openssh \
+    pigz \
+    shadow-uidmap \
+    xz
+
+# alpine-cache-local is the stage to collect apk cache files.
+# This stage can be pushed for sake of reproducible builds.
+FROM scratch AS alpine-cache-local
+COPY --from=alpine-cache-local-base /etc/apk/cache /etc/apk/cache
```

Comment:
Eventually we want to "unsquash" the `/etc/apk/cache` layer to multiple layers in the apk file granularity to help deduplicating the layer blobs:
- https://github.com/moby/buildkit/issues/4236